### ISSUE TITLE
collectd memory: huge pages counted as used

### ISSUE THREAD

**Issue by f1-outsourcing** [collectd/collectd#2914]
Huge pages are counted as used memory, also eg in top. Would be nice to have these as free or available if they are not allocated. What is best way to stat 'real used' memory

```
[@c04 ~]# hugeadm --page-sizes-all
2097152
1073741824
[@c04 ~]# hugeadm --pool-list
      Size  Minimum  Current  Maximum  Default
   2097152     4096     4096     4096        *
1073741824       28       28       28

```

**Comment by octo** [comment]
Could you provide `/proc/meminfo` of an affected system? (Along with some tool output so we can match the content of `/proc/meminfo` to what the user expects.) Thanks!

**Comment by rpv-tomsk** [comment]
I'm not used huge pages yet, but at first look, when huge pages memory pool is configured, then all that memory becomes inaccessible to rest of the system (Is this really true? I'm not sure.). So, I think, reported metrics (and top) are correct.

Maybe we had to add new metrics to report huge pages usage, without touching existing metrics?

**Comment by f1-outsourcing** [comment]
> then all that memory becomes inaccessible to rest of the system (Is this really true? I'm not sure.)

Yes and no, your programs just need to address these pages, if they do not specifically address it, it is not accessible. 

> Maybe we need to add new metrics to report huge pages usage, without touching existing metrics?

I think that is also better

**Comment by sunkuranganath** [comment]
Sorry about the mess up about labels.

@f1-outsourcing this article should provide you more details on how huge page count can be obtained:  https://access.redhat.com/solutions/320303 

Dont think memory pluginis the right place to count number of huge pages.
Have you looked at hugepage plugin of Collectd?

**Comment by sunkuranganath** [comment]
@f1-outsourcing any response? Updating the label.
Once the waiting time for 'waiting for response - 3rd time' expires, this issue will be closed.

**Comment by octo** [comment]
If I understand this issue correctly, than as soon as you configure hugepages, that amount of physical memory is reported as "used" regardless of whether the hugepages are in use or not. This is likely due to the fact that the hugepages are not available to normal memory allocation, i.e. from the point of view of a process *not* using hugepages, that block of memory might as well be in use.

Chances are, though, that the application using hugepages is the one you're most interested in. So in that case, the data reported by the *memory plugin* becomes misleading / useless.

I think we have a couple of options to consider:

*   Subtract `HugePages_Total` from "used" and point users to the *hugepages plugin*.
    Con: the *memory plugin* is supposed to report physical memory, which doesn't disappear due to system configuration.
*   Subtract `HugePages_Total` from "used" and report a "hugepages" metric instead. Point users to the *hugepages plugin* for details.
    Con: feels like artificially limiting the plugin's functionality.
*   Subtract `HugePages_Free` from "used" and add it to "free".
    Con: mixes information about hugepages and "normal" pages.
*   Subtract `HugePages_Total` from "used" and report used/free/… hugepages as separate metrics.
    Con: mostly obsoletes the *hugepages plugin*.

None of these options strike me as prohibitively complicated to implement. Which one would provide the most use for users?

Best regards,
—octo

**Comment by sunkuranganath** [comment]
Any time hugpages are needed on a system, the exact number of hugepages with a specific size (2MB or 1GB, etc.) need to be set in the BIOS at the time of system startup. This means, the administrator need to account for them right from the beginning while the memory stats considers them as used.
So, IMHO, none of the above 4 mentioned options might justify hugepage count and hugepage plugin. It might be better to indicate in memory plugin documentation that configured hugepage count is considered used and hugepage plugin is necessary to better figure out amount of used/reserved/free hugepages. This way we can keep both of them separate.
